src/main/java/com/conveyal/gtfs/validator/FlexValidator.java

```diff
@@ -0,0 +1,358 @@
+package com.conveyal.gtfs.validator;
+
+import com.conveyal.gtfs.error.NewGTFSError;
+import com.conveyal.gtfs.error.NewGTFSErrorType;
+import com.conveyal.gtfs.error.SQLErrorStorage;
+import com.conveyal.gtfs.loader.Feed;
+import com.conveyal.gtfs.model.BookingRule;
+import com.conveyal.gtfs.model.FareRule;
+import com.conveyal.gtfs.model.Location;
+import com.conveyal.gtfs.model.LocationGroup;
+import com.conveyal.gtfs.model.Stop;
+import com.conveyal.gtfs.model.StopTime;
+import com.google.common.collect.Lists;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.conveyal.gtfs.model.Entity.DOUBLE_MISSING;
+import static com.conveyal.gtfs.model.Entity.INT_MISSING;
+import static com.conveyal.gtfs.util.GeoJsonUtil.GEOMETRY_TYPE_POLYGON;
+
+/**
+ * Spec validation checks for flex additions as defined here:
+ * https://github.com/MobilityData/gtfs-flex/blob/master/spec/reference.md
+ *
+ * Number of checks:
+ *
+ * Location group: 2
+ * Location: 2
+ * Stop times: 14
+ * Booking rules: 10
+ */
+public class FlexValidator extends FeedValidator {
+
+    // Store all validation errors here so that they can be checked as part of testing.
+    public final List<NewGTFSError> errors = new ArrayList<>();
+
+    /**
+     * Constructor for testing.
+     */
+    public FlexValidator() {
+        this(null, null);
+    }
+
+    public FlexValidator(Feed feed, SQLErrorStorage errorStorage) {
+        super(feed, errorStorage);
+    }
+
+    @Override
+    public void validate() {
+        feed.locationGroups.forEach(locationGroup ->
+            validateLocationGroups(locationGroup, Lists.newArrayList(feed.stops), Lists.newArrayList(feed.locations))
+        );
+        feed.locations.forEach(location ->
+            validateLocations(location, Lists.newArrayList(feed.stops), Lists.newArrayList(feed.fareRules))
+        );
+        feed.stopTimes.forEach(stopTime ->
+            validateStopTimes(stopTime, Lists.newArrayList(feed.locationGroups), Lists.newArrayList(feed.locations))
+        );
+        feed.bookingRules.forEach(this::validateBookingRules);
+        // Register errors, if any, once all checks have been completed.
+        errors.forEach(this::registerError);
+    }
+
+    public void validateLocationGroups(LocationGroup locationGroup, List<Stop> stops, List<Location> locations) {
+        if (!stops.isEmpty() && stops.stream().anyMatch(stop -> stop.stop_id.equals(locationGroup.location_group_id))) {
```

Comment:
If not passing fields from `feed`, could something like `StreamSupport.stream(feed.stops.spliterator(), false)` work, or would it be too clunky?

Reply:
This `if` condition is important logic for the validator, it should be extracted to a method `locationGroupIdIsStopId(String locationGroupId)` that utilizes `feed.stops` (or a more generic version of that) and lets you delete the comment below.

Same applies at least to the other checks in this class that use `anyMatch`.